core/src/main/scala/org/apache/spark/executor/ExecutorURLClassLoader.scala

```diff
@@ -32,36 +36,52 @@ private[spark] trait MutableURLClassLoader extends ClassLoader {
 }
 
 private[spark] class ChildExecutorURLClassLoader(urls: Array[URL], parent: ClassLoader)
-  extends MutableURLClassLoader {
+  extends URLClassLoader(urls, null) with MutableURLClassLoader {
+
+  private val parentClassLoader = new ParentClassLoader(parent)
+  private val locks = new ConcurrentHashMap[String, Object]()
 
-  private object userClassLoader extends URLClassLoader(urls, null){
-    override def addURL(url: URL) {
-      super.addURL(url)
+  override def loadClass(name: String, resolve: Boolean): Class[_] = {
+    var lock = locks.get(name)
+    if (lock == null) {
+      val newLock = new Object()
+      lock = locks.putIfAbsent(name, newLock)
+      if (lock == null) {
+        lock = newLock
+      }
     }
-    override def findClass(name: String): Class[_] = {
-      super.findClass(name)
+
+    lock.synchronized {
+      try {
+        super.loadClass(name, resolve)
+      } catch {
+        case e: ClassNotFoundException =>
+          parentClassLoader.loadClass(name, resolve)
+      }
     }
   }
 
-  private val parentClassLoader = new ParentClassLoader(parent)
+  override def getResource(name: String): URL = {
+    val url = super.findResource(name)
+    val res = if (url != null) url else parentClassLoader.getResource(name)
+    res
+  }
 
-  override def findClass(name: String): Class[_] = {
-    try {
-      userClassLoader.findClass(name)
-    } catch {
-      case e: ClassNotFoundException => {
-        parentClassLoader.loadClass(name)
+  override def getResources(name: String): Enumeration[URL] = {
+    val urls = super.findResources(name)
+    val res =
+      if (urls != null && urls.hasMoreElements()) {
+        urls
+      } else {
+        parentClassLoader.getResources(name)
       }
-    }
+    res
   }
 
-  def addURL(url: URL) {
-    userClassLoader.addURL(url)
+  override def addURL(url: URL) {
+    super.addURL(url)
   }
 
-  def getURLs() = {
-    userClassLoader.getURLs()
-  }
 }
 
 private[spark] class ExecutorURLClassLoader(urls: Array[URL], parent: ClassLoader)
```

Comment:
This shouldn't be called ExecutorURLClassLoader if it's being used to load things in the driver.